### PR TITLE
SQL: Preserve original source for each expression

### LINKED ISSUE #36894: SQL: Preserve original source for each expression
Currently, after each expression is parsed the original query statement is not tracked anymore. While this works it creates an issue when displaying an expression that is not named since the reconstructed string can and will vary from the original declaration (upper/lower casing, extra spaces, etc...).

For consistency, just like with location, the initial text should be preserved.

### PR DESCRIPTION
**Pull request by costin**
The source text is saved alongside the location of each Node/Expression allowing accurate reproducibility of the source

Fix #36894

The PR is small but it is wide spread hence the huge number of files involved.
I'd like to add more tests to check the source is being properly preserved. 

A follow-up will make sure the source is displayed as is for naming.

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertFunctionPipe.java [hunk 1]
```diff
@@ -90,7 +90,7 @@ public InsertFunctionProcessor asProcessor() {
         return new InsertFunctionProcessor(source.asProcessor(), start.asProcessor(), length.asProcessor(), replacement.asProcessor());
     }
     
-    public Pipe source() {
+    public Pipe sourcePipe() {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/InsertFunctionPipe.java]: Maybe change the name of the rest of the methods to the `*Pipe()` format? It doesn't look consistent... `source` has its method called `sourcePipe()` while `start` and the rest have their methods called `start()`, `length()`, `replacement()`. Or maybe change the name of the `source` field to `sourceText` and have its return method called `sourceText()`. I would be in favor of the second option, to remain consistent with the rest of the functions (that don't have `Pipe` in their return methods).

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/LocateFunctionPipe.java [hunk 2]
```diff
@@ -82,7 +82,7 @@ public LocateFunctionProcessor asProcessor() {
         return new LocateFunctionProcessor(pattern.asProcessor(), source.asProcessor(), start == null ? null : start.asProcessor());
     }
     
-    public Pipe source() {
+    public Pipe sourcePipe() {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/LocateFunctionPipe.java]: Same comment here as the one for `Insert` function: maybe change `source` to `sourceText` and don't use `Pipe` in the return method name.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/ReplaceFunctionPipe.java [hunk 3]
```diff
@@ -78,7 +78,7 @@ public ReplaceFunctionProcessor asProcessor() {
         return new ReplaceFunctionProcessor(source.asProcessor(), pattern.asProcessor(), replacement.asProcessor());
     }
     
-    public Pipe source() {
+    public Pipe sourcePipe() {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/ReplaceFunctionPipe.java]: Here, as well.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/SubstringFunctionPipe.java [hunk 4]
```diff
@@ -78,7 +78,7 @@ public SubstringFunctionProcessor asProcessor() {
         return new SubstringFunctionProcessor(source.asProcessor(), start.asProcessor(), length.asProcessor());
     }
     
-    public Pipe source() {
+    public Pipe sourcePipe() {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/SubstringFunctionPipe.java]: Here

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/execution/search/SourceGeneratorTests.java [hunk 5]
```diff
@@ -95,13 +94,13 @@ public void testSortFieldSpecified() {
         FieldSortBuilder sortField = fieldSort("test").unmappedType("keyword");
         
         QueryContainer container = new QueryContainer()
-                .sort(new AttributeSort(new FieldAttribute(new Location(1, 1), "test", new KeywordEsField("test")), Direction.ASC,
+                .sort(new AttributeSort(new FieldAttribute(Source.EMPTY, "test", new KeywordEsField("test")), Direction.ASC,
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/execution/search/SourceGeneratorTests.java]: Here you replaced `new Location(1,1)` with `Source.EMPTY`, but [here](https://github.com/elastic/elasticsearch/pull/36912/files#diff-c555d1468ca62f7fbc3f7ec9eb513e79R24) you had a different approach.

**Review by astefan — APPROVED** [review]
LGTM.

Left few comments regarding the string manipulating functions that have one of their parameters called `source`: REPLACE(), INSERT(), LOCATE() and SUBSTRING(). I was suggesting in my comments to have the `source` parameter name changed to something like `sourceText` and not changing the return method name to `sourcePipe()`.

**Review by matriv — APPROVED** [review]
LGTM. I agree with Andrei's comments about the `sourcePipe`.
On top I'd suggest to change the methods' parameter name from `location` to `source`, e.g.: https://github.com/elastic/elasticsearch/pull/36912/files#diff-f90669a222551c32dc8c7fdf880eba57R68 This will create a few conflicts as some expression params are names `source` which can be renamed to something else like `sourceStr` or `sourceText` as Andrei said.

**Comment by costin** [comment]
run the gradle build tests 1

**Pull request merged**